README.md
# dircast [![Build Status](https://travis-ci.org/frodeaa/dircast.svg)](https://travis-ci.org/frodeaa/dircast)

Simple tool to generate RSS feed
from mp3 files found in a directory

    usage: dircast [<flags>] <directory>

    Flags:
      --help             Show help.
      -s, --server=http://localhost:8000/
                         hostname (and path) to the root e.g. http://myserver.com/rss
      -b, --bind         Start HTTP server, bind to the server
      --log              Enable log of HTTP requests
      -r, --recursive    how to handle the directory scan
      -a, --auto-image   Resolve RSS image automatically, will use ID3 attached
                         image, image overrides this option, only available in
                         combination with bind
      -l, --language=LANGUAGE
                         the language of the RSS document, a ISO 639 value
      -t, --title="RSS FEED"
                         RSS channel title
      -d, --description=DESCRIPTION
                         RSS channel description
      -i, --image=IMAGE  Image URL for the RSS channel image
      -f, --file="mp3"   File type to include in the RSS document
      --version          Show application version.

    Args:
      <directory>  directory to read files relative from

## Build and test

    make
    bin/dircast --help

## Install

    make
    sudo make install


CHANGELOG.md
## 0.4.0 (2016-05-21)

Features:

 - adds a `--log` to enable log if HTTP request

Bugs:

 - fix trailing nulls read from ID3 tag
 - fix Makefile, build would fail if GOPATH was initial empty

## 0.3.0 (2015-03-29)

Features:

 - adds a `--auto-image` to use ID3 attached image as RSS Channel image

Bugs:

 - Fix item URL path, the path would contain `//`, e.g. http://server:port//media.mp3

## 0.2.0 (2015-02-24)

Features:

  - adds a `--bind` flag to run a HTTP server hosting the RSS feed
  - adds a `--file` parameter to support custome file types (default is mp3)


dircast_test.go
package main

import (
	"os"
	"strings"
	"testing"
)

func TestFormatYear(t *testing.T) {
	cases := []struct {
		in, want string
	}{
		{"2015", "Thu, 01 Jan 2015 00:00:00 +0000"},
		{"20151102", "Mon, 02 Nov 2015 00:00:00 +0000"},
		{"20152402", "Tue, 24 Feb 2015 00:00:00 +0000"},
		{"", ""},
		{"NOT_A_YEAR", ""},
	}
	for _, c := range cases {
		got := formatYear(c.in)
		if got != c.want {
			t.Errorf("formatYear(%q) == %q, want %q", c.in, got, c.want)
		}
	}
}

func TestFileUrl(t *testing.T) {
	cases := []struct {
		relativePath, baseUrl, want string
	}{
		{"", "", ""},
		{"c/d/e", "a/b/", "a/b/c/d/e"},
		{"c d", "a/b/", "a/b/c%20d"},
		{"/c", "a/b/", "a/b/c"},
	}
	for _, c := range cases {
		got := fileUrl(c.relativePath, c.baseUrl)
		if got != c.want {
			t.Errorf("fileUrl(%q, %q) == %q, want %q", c.relativePath, c.baseUrl, got, c.want)
		}
	}
}

func TestAddMetaTitle(t *testing.T) {
	title := "Nice Life (Feat. Basick)"
	in := Item{}
	path := "./vendor/src/github.com/mikkyang/id3-go/test.mp3"
	f, _ := os.Stat(path)

	addMeta(path, f, &in, false)

	if in.Title != title {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.Title == %q want %q", in.Title, title)
	}
}

func TestAddMetaTitleFallbackOnFilieName(t *testing.T) {
	title := "test.mp3"
	in := Item{}
	f, _ := os.Stat("./vendor/src/github.com/mikkyang/id3-go/test.mp3")

	addMeta("NOT_PATH_TO_FILE", f, &in, false)

	if in.Title != title {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.Title == %q want %q", in.Title, title)
	}
}

func TestAddMetaPubDate(t *testing.T) {
	pubDate := "Mon, 25 Nov 2013 00:00:00 +0000"
	in := Item{}
	path := "./vendor/src/github.com/mikkyang/id3-go/test.mp3"
	f, _ := os.Stat(path)

	addMeta(path, f, &in, false)

	if in.PubDate != pubDate {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.PubDate == %q want %q", in.PubDate, pubDate)
	}

}

func TestAddMetaSubTitle(t *testing.T) {
	subtitle := "Paloalto"
	in := Item{}
	path := "./vendor/src/github.com/mikkyang/id3-go/test.mp3"
	f, _ := os.Stat(path)

	addMeta(path, f, &in, false)

	if !strings.HasPrefix(in.Subtitle, subtitle) {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.Subtitle ~= %q want %q", in.Subtitle, subtitle)
	}

}

func TestVisitFiles(t *testing.T) {

	cases := []struct {
		path, fileType string
		want           int
	}{
		{"./vendor/src/github.com/mikkyang/id3-go/", "mp3", 0},
		{"./vendor/src/github.com/mikkyang/id3-go/test.mp3", "mp3", 1},
	}

	for _, c := range cases {
		channel := &Channel{}
		f, e := os.Stat(c.path)
		v := visitFiles(c.path, channel, "test://", false, c.fileType, false)
		v(c.path, f, e)
		if len(channel.Items) != c.want {
			t.Errorf("visitFiles(%q, channel,...), len(channel.Items) == %d want %d",
				c.path, len(channel.Items), c.want)
		}
	}

}


src/github.com/frodeaa/dircast/dircast_test.go
package main

import (
	"fmt"
	rss "github.com/frodeaa/dircast/core"
	"os"
	"strings"
	"testing"
)

func TestFormatYear(t *testing.T) {
	cases := []struct {
		in, want string
	}{
		{"2015", "Thu, 01 Jan 2015 00:00:00 +0000"},
		{"20151102", "Mon, 02 Nov 2015 00:00:00 +0000"},
		{"20152402", "Tue, 24 Feb 2015 00:00:00 +0000"},
		{"", ""},
		{"NOT_A_YEAR", ""},
	}
	for _, c := range cases {
		got := formatYear(c.in)
		if got != c.want {
			t.Errorf("formatYear(%q) == %q, want %q", c.in, got, c.want)
		}
	}
}

func TestFileUrl(t *testing.T) {
	cases := []struct {
		relativePath, baseUrl, want string
	}{
		{"", "", ""},
		{"c/d/e", "a/b/", "a/b/c/d/e"},
		{"c d", "a/b/", "a/b/c%20d"},
		{"/c", "a/b/", "a/b/c"},
	}
	for _, c := range cases {
		got := fileUrl(c.relativePath, c.baseUrl)
		if got != c.want {
			t.Errorf("fileUrl(%q, %q) == %q, want %q", c.relativePath, c.baseUrl, got, c.want)
		}
	}
}

func TestAddMetaTitle(t *testing.T) {
	title := "Nice Life (Feat. Basick)"
	in := rss.Item{}
	path := "./../../../../vendor/src/github.com/mikkyang/id3-go/test.mp3"
	f, _ := os.Stat(path)

	addMeta(path, f, &in, false)

	if in.Title != title {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.Title == %q want %q", in.Title, title)
	}
}

func TestAddMetaTitleFallbackOnFilieName(t *testing.T) {
	title := "test.mp3"
	in := rss.Item{}
	f, e := os.Stat("./../../../../vendor/src/github.com/mikkyang/id3-go/test.mp3")
	if e != nil {
		fmt.Errorf(" A %s", e)
	}

	addMeta("NOT_PATH_TO_FILE", f, &in, false)

	if in.Title != title {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.Title == %q want %q", in.Title, title)
	}
}

func TestAddMetaPubDate(t *testing.T) {
	pubDate := "Mon, 25 Nov 2013 00:00:00 +0000"
	in := rss.Item{}
	path := "./../../../../vendor/src/github.com/mikkyang/id3-go/test.mp3"
	f, _ := os.Stat(path)

	addMeta(path, f, &in, false)

	if in.PubDate != pubDate {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.PubDate == %q want %q", in.PubDate, pubDate)
	}

}

func TestAddMetaSubTitle(t *testing.T) {
	subtitle := "Paloalto"
	in := rss.Item{}
	path := "./../../../../vendor/src/github.com/mikkyang/id3-go/test.mp3"
	f, _ := os.Stat(path)

	addMeta(path, f, &in, false)

	if !strings.HasPrefix(in.Subtitle, subtitle) {
		t.Errorf("addMeta(<test.mp3>, <FileInfo>, <Item>) Item.Subtitle ~= %q want %q", in.Subtitle, subtitle)
	}

}

func TestVisitFiles(t *testing.T) {

	cases := []struct {
		path, fileType string
		want           int
	}{
		{"./../../../../vendor/src/github.com/mikkyang/id3-go/", "mp3", 0},
		{"./../../../../vendor/src/github.com/mikkyang/id3-go/test.mp3", "mp3", 1},
	}

	for _, c := range cases {
		channel := &rss.Channel{}
		f, e := os.Stat(c.path)
		v := visitFiles(c.path, channel, "test://", false, c.fileType, false)
		v(c.path, f, e)
		if len(channel.Items) != c.want {
			t.Errorf("visitFiles(%q, channel,...), len(channel.Items) == %d want %d",
				c.path, len(channel.Items), c.want)
		}
	}

}


src/github.com/frodeaa/dircast/core/rss_test.go
package core

import (
	"testing"
)

type Output struct {
	Writer
	data string
}

func (o *Output) Write(data []byte) (int, error) {
	o.data += string(data)
	return 0, nil
}

func TestOut(t *testing.T) {
	channel := Channel{
		PubDate:     "PUBDATE",
		Title:       "TITLE",
		Link:        "LINK",
		Description: "DESCRIPTION",
		Language:    "LANGUAGE",
	}

	channel.Items =
		append(channel.Items, Item{})

	out := &Output{}
	(&Rss{
		Channel: channel,
		Version: "2.0",
		NS:      "itunes",
	}).Out(out)

	want := "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
		"\n<rss version=\"2.0\" xmlns:itunes=\"itunes\">" +
		"\n  <channel>\n    <pubDate>PUBDATE</pubDate>" +
		"\n    <title>TITLE</title>\n    <link>LINK</link>" +
		"\n    <description>DESCRIPTION</description>" +
		"\n    <language>LANGUAGE</language>\n    <item>" +
		"\n      <title></title>\n      <description></description>" +
		"\n      <enclosure url=\"\" length=\"0\" type=\"\"></enclosure>" +
		"\n      <guid></guid>\n    </item>\n  </channel>\n</rss>"

	if out.data != want {
		t.Errorf("rss.Out(writer) = %q, want %q", out.data, want)
	}

}


src/github.com/frodeaa/dircast/core/rss.go
package core

import "encoding/xml"

const (
	Header   = `<?xml version="1.0" encoding="UTF-8"?>` + "\n"
	ITunesNs = "http://www.itunes.com/dtds/podcast-1.0.dtd"
)

type Rss struct {
	XMLName xml.Name `xml:"rss"`
	Channel Channel  `xml:"channel"`
	Version string   `xml:"version,attr"`
	NS      string   `xml:"xmlns:itunes,attr"`
}

type Channel struct {
	PubDate     string  `xml:"pubDate,omitempty"`
	Title       string  `xml:"title,omitempty"`
	Link        string  `xml:"link,omitempty"`
	Description string  `xml:"description"`
	Language    string  `xml:"language,omitempty"`
	Images      []Image `xml:"image,omitempty"`
	Items       []Item  `xml:"item"`
}

type Item struct {
	Title       string    `xml:"title"`
	Description string    `xml:"description"`
	Enclosure   Enclosure `xml:"enclosure"`
	Guid        string    `xml:"guid"`
	Subtitle    string    `xml:"itunes:subtitle,omitempty"`
	Categories  []Text    `xml:"itunes:category,omitempty"`
	PubDate     string    `xml:"pubDate,omitempty"`
}

type Text struct {
	Value string `xml:"text,attr"`
}

type Image struct {
	Link  string `xml:"link"`
	Title string `xml:"title"`
	Url   string `xml:"url"`
}

type Enclosure struct {
	Url    string `xml:"url,attr"`
	Length int64  `xml:"length,attr"`
	Type   string `xml:"type,attr"`
}

type Writer interface {
	Write([]byte) (int, error)
}

func (rss *Rss) Out(w Writer) error {
	body, err := xml.MarshalIndent(rss, "", "  ")
	w.Write([]byte(Header))
	w.Write([]byte(body))
	return err
}


src/github.com/frodeaa/dircast/core/server.go
package core

import (
	"log"
	"net/http"
	"net/url"
	"strings"
)

type rssHandler struct {
	feed      Rss
	fs        http.Handler
	path      string
	imageBlob []byte
}

func NewRssHandler(source Source) *rssHandler {
	return &rssHandler{feed: *source.Rss(),
		fs: http.FileServer(http.Dir(source.Root)), imageBlob: source.image}
}

func Log(handler http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Printf("%s %s %s", r.RemoteAddr, r.Method, r.URL)
		handler.ServeHTTP(w, r)
	})
}

func isImagePath(path string, images []Image) bool {
	return len(images) > 0 && strings.HasSuffix(images[0].Url, path)
}

func (rss *rssHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path
	if path == "" || path == "/" {
		rss.feed.Out(w)
	} else if len(rss.imageBlob) > 0 && isImagePath(path, rss.feed.Channel.Images) {
		w.Write(rss.imageBlob)
	} else {
		http.StripPrefix(rss.path, rss.fs).ServeHTTP(w, r)
	}
}

func contentPath(url *url.URL) string {
	path := url.Path
	if !strings.HasSuffix(path, "/") {
		path += "/"
	}
	return path
}

func Server(source Source, logEnabled bool) error {

	url, err := url.Parse(source.PublicUrl)

	if err != nil {
		return err
	}

	http.Handle(contentPath(url), NewRssHandler(source))
	if logEnabled {
		http.ListenAndServe(url.Host, Log(http.DefaultServeMux))
	}
	return http.ListenAndServe(url.Host, nil)

}


src/github.com/frodeaa/dircast/core/source.go
package core

import (
	"crypto/sha1"
	"encoding/base64"
	id3 "github.com/mikkyang/id3-go"
	"github.com/mikkyang/id3-go/v2"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

type Source struct {
	channel   Channel
	Root      string
	PublicUrl string
	recursive bool
	fileType  string
	autoImage bool
	image     []byte
}

type MediaFile struct {
	*id3.File
	defaultName string
}

type MediaItem struct {
	Item
}

func trimmed(value string) string {
	cutset := string(rune(0))
	return strings.Trim(strings.TrimRight(value, cutset), " ")
}

func fileUrl(relativePath string, baseUrl string) string {
	Url, _ := url.Parse(baseUrl)
	Url.Path += relativePath
	Url.Path = strings.Replace(Url.Path, "//", "/", -1)
	return Url.String()
}

func formatYearRFC1123Z(pattern, year string) (string, error) {
	t, err := time.Parse(pattern, year)
	if err != nil {
		return "", err
	}
	return t.Format(time.RFC1123Z), nil
}

func formatYear(year string) string {
	if len(year) > 0 {
		t, err := formatYearRFC1123Z("20060102", year)
		if err == nil {
			return t
		}
		t, err = formatYearRFC1123Z("2006", year)
		if err == nil {
			return t
		}
		t, err = formatYearRFC1123Z("20060201", year)
		if err == nil {
			return t
		}
	}
	return ""
}

func (fd *MediaFile) titleArtist() string {
	title := trimmed(fd.Title())
	author := trimmed(fd.Artist())
	var res string
	if len(title) > 0 {
		res = title
	} else {
		res = author
		if len(author) > 0 {
			res += " - "
		}
		res += fd.defaultName
	}
	return res
}

func (fd *MediaFile) readImage() []byte {
	var data []byte
	apic := fd.Frame("APIC")
	if apic != nil {
		switch t := apic.(type) {
		case *v2.ImageFrame:
			data = v2.ImageFrame(*t).Data()
		}
	}
	return data
}

func (fd *MediaFile) copyCategoryTo(m *MediaItem) {
	tcon := fd.Frame("TCON")
	if tcon != nil {
		m.Item.Categories = append(m.Item.Categories, Text{Value: trimmed(tcon.String())})
	}
}

func (fd *MediaFile) yearFormatted() string {
	return formatYear(fd.Year())
}

func (fd *MediaFile) copyMetaTo(m *MediaItem) {
	m.Title = fd.titleArtist()
	m.Item.PubDate = fd.yearFormatted()
	fd.copyCategoryTo(m)
}

func (m *MediaItem) addMeta(path, defaultName string, source *Source) {
	fd, err := id3.Open(path)
	if err != nil {
		m.Item.Title = defaultName
	} else {
		media := &MediaFile{fd, defaultName}
		defer fd.Close()
		media.copyMetaTo(m)
		if source.autoImage && len(source.image) == 0 {
			source.SetImage(media.readImage())
		}
	}
}

func NewSource(root string, recursive bool, publicUrl,
	title, description, language, filetype string, autoImage bool) *Source {
	channel := &Channel{
		PubDate:     time.Now().Format(time.RFC1123Z),
		Title:       title,
		Link:        publicUrl,
		Description: description,
		Language:    language}
	return &Source{
		Root:      root,
		recursive: recursive,
		PublicUrl: publicUrl,
		channel:   *channel,
		fileType:  filetype,
		autoImage: autoImage}
}

func (s *Source) SetImage(image []byte) {
	s.image = image
	if len(image) > 0 {
		hasher := sha1.New()
		hasher.Write(image)
		s.channel.Images = append(s.channel.Images,
			Image{Title: s.channel.Title, Link: s.channel.Link, Url: s.channel.Link + base64.URLEncoding.EncodeToString(hasher.Sum(nil))})
	}
}

func (s *Source) SetChannelImageUrl(url string) {
	s.channel.Images = append(s.channel.Images, Image{Title: s.channel.Title, Link: s.channel.Link, Url: url})
}

func (s *Source) addFile(path string, info os.FileInfo) {
	url := fileUrl(path[len(s.Root)-1:], s.PublicUrl)
	item := MediaItem{Item{Enclosure: Enclosure{Length: info.Size(), Type: "audio/mpeg",
		Url: url}, Guid: url}}
	item.addMeta(path, info.Name(), s)
	s.channel.Items = append(s.channel.Items, item.Item)
}

func (s *Source) HandleWalk() filepath.WalkFunc {
	return func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		if !s.recursive && path != s.Root && info.IsDir() {
			return filepath.SkipDir
		}

		if !!info.IsDir() {
			return nil
		}

		matched, _ := filepath.Match("*."+s.fileType, info.Name())
		if matched {
			s.addFile(path, info)
		}
		return nil
	}
}

func (s *Source) Rss() *Rss {
	return &Rss{Channel: s.channel, Version: "2.0", NS: ITunesNs}
}


src/github.com/frodeaa/dircast/dircast.go
package main

import (
	"fmt"
	dircast "github.com/frodeaa/dircast/core"
	kingpin "gopkg.in/alecthomas/kingpin.v1"
	"os"
	"os/signal"
	"path/filepath"
	"strings"
	"syscall"
)

var (
	baseUrl     = kingpin.Flag("server", "hostname (and path) to the root e.g. http://myserver.com/rss").Short('s').Default("http://localhost:8000/").URL()
	bind        = kingpin.Flag("bind", "Start HTTP server, bind to the server").Short('b').Bool()
	logEnabled  = kingpin.Flag("log", "Enable log of HTTP requests").Bool()
	recursive   = kingpin.Flag("recursive", "how to handle the directory scan").Short('r').Bool()
	autoImage   = kingpin.Flag("auto-image", "Resolve RSS image automatically, will use ID3 attached image, image overrides this option, only available in combination with bind").Short('a').Bool()
	language    = kingpin.Flag("language", "the language of the RSS document, examples: 'no' 'en-us', (ISO 639)").Short('l').String()
	title       = kingpin.Flag("title", "RSS channel title").Short('t').Default("RSS FEED").String()
	description = kingpin.Flag("description", "RSS channel description").Short('d').String()
	imageUrl    = kingpin.Flag("image", "Image URL for the RSS channel image").Short('i').URL()
	fileType    = kingpin.Flag("file", "File type to include in the RSS document").Short('f').Default("mp3").String()
	path        = kingpin.Arg("directory", "directory to read files relative from").Required().ExistingDir()
)

func writeStartupMsg(workdir string, url string) {
	fmt.Printf(
		"\x1b[33;1m%v\x1b[0m \x1b[36;1m%v\x1b[0m \x1b[33;1mon:\x1b[0m \x1b[36;1m%v\x1b[0m\n",
		"Starting up dircast, serving", workdir, url)
	fmt.Println("Hit CTRL-C to stop the server")
}

func onShutdown(message string) {
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	signal.Notify(c, syscall.SIGTERM)
	go func() {
		<-c
		fmt.Printf("\x1b[31;1m%v\x1b[0m\n", message)
		os.Exit(1)
	}()
}

func main() {

	kingpin.Version("0.4.0")
	kingpin.Parse()

	if !strings.HasSuffix((*baseUrl).Path, "/") {
		(*baseUrl).Path = (*baseUrl).Path + "/"
	}

	if !*bind || *imageUrl != nil {
		*autoImage = false
	}

	source := dircast.NewSource(*path, *recursive, (*baseUrl).String(), *title, *description, *language, *fileType, *autoImage)
	if *imageUrl != nil {
		source.SetChannelImageUrl((*imageUrl).String())
	}
	err := filepath.Walk(*path, source.HandleWalk())

	if err != nil {
		fmt.Printf("%s: %v\n", os.Args[0], err)
	} else {
		if *bind {
			onShutdown("dircast stopped.")
			writeStartupMsg(source.Root, source.PublicUrl)
			err = dircast.Server(*source, *logEnabled)
			if err != nil {
				fmt.Printf("error: %v\n", err)
			}
		} else {
			source.Rss().Out(os.Stdout)
		}
	}

}
